docs/adrs/0009-nimbus-web-sdk-architecture.md

```diff
@@ -0,0 +1,97 @@
+The `nimbus-sdk` and `cirrus-sdk` libraries would depend on the `nimbus-core` library, extending the functionality for their various use cases. 
+
+* Good, because separate libraries allow for strong separation of concerns.
+* Bad, because defining and using types becomes more difficult.
+  * Traits will be needed to extend functionality in the Nimbus Core library. While dynamic typing with traits is supported, it is one of the more complex features of Rust.
+* Bad, because it would result in a significant amount of code churn.
+
+### Cargo features, one library
+
+Cargo features allow for conditional compilation of libraries and library dependencies. 
+Native Rust attributes are applied to sections of Rust code to conditionally compile structs, methods, variables, and more.
+
+For more information, check out the Cargo Features reference[<sup>[1.ii]</sup>](#links) available in the Rust Cargo book.
+
+* Good, because the use of feature configurations is familiar to Rust developers.
+* Good, because all code relating to Nimbus business logic remains in one place.
+* Good, because UniFFI bindings can be conditionally generated for each feature with different UDLs.
+* Good, because dependencies can be excluded based on enabled features.
+  * As an example, RKV will only be needed for Nimbus, so the RKV dependency will only be included in compilation when the `nimbus` feature is enabled.
+* Good, because it will not increase the size of the existing Nimbus SDK binary.
+* Bad, because onboarding developers learning Rust would likely have a more difficult time understanding Cargo features.
+
+### Cargo features, separate library
+
+This is roughly the same concept as above, but the features will be used in a different way.
+Instead of features for both Nimbus and Cirrus, we would include a `stateless` feature in the Nimbus library to disable state-oriented features and dependencies.
+The Cirrus library will then exist as a separate library from the Nimbus library.
+
+* Good, because all the good things in the "Cargo features, one library" option above.
+* Good, because the changes required will require minimal churn.
+* Good, because better separation of concerns between Nimbus and Cirrus
+* Bad, because defining and using types and traits across the libraries could prove difficult.
+  * See a similar point in the "Separate `nimbus-core`, `nimbus-sdk`, and `cirrus-sdk` libraries" option above.
+* Bad, because methods used across both libraries could result in collisions during development.
+
+### Expose a stateless API alongside the stateful API
+
+In this solution, we would expose static/stateless methods for interacting with experiments.
+The UniFFI UDL would be updated to include the methods, resulting in the methods being visible in Android and iOS where they will not be used. 
+
+* Good, because it's simple and entirely additive.
+* Bad, because it makes Nimbus more prone to developer error.
```

Comment:
Can you provide some more context or examples about how this would occur?

Reply:
Sure! So the existing UDL includes a NimbusClient class – this class shouldn't ever actually be used by iOS/Android directly, but it is there and _can_ be used. It doesn't include any tooling for handling threading or object builders or anything like that.

By adding the stateless methods for experiment evaluation (and others), they would be available in Python, Kotlin, and Swift, even though they should only be used within the context of Cirrus. 

More than anything else I see this as a "someone finds the function in the Nimbus internals and starts using it incorrectly" concern.

Reply:
Ah interesting.  That makes sense, we should also add some of that context back to the ADR.